### PR TITLE
GenAI servables - refactor input processing

### PR DESCRIPTION
**Pull request by mzegla**

### CODE REVIEW

File: src/llm/io_processing/input_processor_context.hpp [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace ovms {
+
+// Holds the per-deployment resources needed by InputProcessor.
+// Created once during servable initialization; reused across requests.
+struct InputProcessorContext {
+    InputProcessingConfig config;
+    ov::genai::Tokenizer tokenizer;
+#if (PYTHON_DISABLE == 0)
+    PyJinjaTemplateProcessor* templateProcessor = nullptr;
```
[the remark this answers is not in the document]
Comment (mzegla) [reply src/llm/io_processing/input_processor_context.hpp:35]: tokenizer type is only a shared pointer wrapper on tokenizer impl
so effectively it already is a pointer

File: src/llm/servable.hpp [hunk 2]
```diff
@@ -36,7 +36,9 @@
 #include "../http_payload.hpp"
 #include "../sse_utils.hpp"
 #include "apis/openai_api_handler.hpp"
-#include "io_processing/generation_config_builder.hpp"
+#include "io_processing/base_generation_config_builder.hpp"
+#include "io_processing/input_processor_context.hpp"
+#include "io_processing/input_request.hpp"
```
[the remark this answers is not in the document]
Comment (mzegla) [reply src/llm/servable.hpp:41]: It is needed for DecodingMethod

File: src/llm/apis/openai_api_handler.cpp [hunk 3]
```diff
@@ -492,18 +483,48 @@ const OpenAIRequest& OpenAIApiHandler::getRequest() const {
...
+    req.generationConfig = configBuilder.getConfig();
+    if (endpoint == Endpoint::COMPLETIONS) {
+        req.input = request.prompt.value_or("");
+    } else {
+        // CHAT_COMPLETIONS and RESPONSES both use ChatHistory.
+        // Copied (not moved) so the handler retains its own copy for response serialization.
+        req.input = request.chatHistory;
+        // Populate tools and chat_template_kwargs on the copied ChatHistory so
+        // ChatTemplateProcessor can access them via get_tools()/get_extra_context().
+        auto& chatHistory = std::get<ov::genai::ChatHistory>(req.input);
```
Comment (dkalinowski) [inline src/llm/apis/openai_api_handler.cpp:509]: it can throw an exception if req.input is not of type ChatHistory. can we try catch that and return error if it happens?

File: src/llm/io_processing/input_processors/chat_template_processor.hpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    // Minja / native-OV path: no PyJinja processor needed.
+    explicit ChatTemplateProcessor(ov::genai::Tokenizer& tokenizer);
+#else
+    explicit ChatTemplateProcessor(ov::genai::Tokenizer& tokenizer);
+#endif
+
+    absl::Status process(InputRequest& req) override;
+
+private:
+    ov::genai::Tokenizer* tokenizer;  // non-owning; lifetime tied to InputProcessorContext
```
Comment (dkalinowski) [inline src/llm/io_processing/input_processors/chat_template_processor.hpp]: nit pick - we can do that since all constructors take tokenizer ref as argument
```suggestion
    ov::genai::Tokenizer& tokenizer;  // non-owning; lifetime tied to InputProcessorContext
```

File: src/llm/io_processing/input_processors/image_decoding_processor.cpp [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+namespace ovms {
+
+ImageDecodingProcessor::ImageDecodingProcessor(
+    std::optional<std::string> allowedLocalMediaPath,
+    std::optional<std::vector<std::string>> allowedMediaDomains) :
+    allowedLocalMediaPath(std::move(allowedLocalMediaPath)),
+    allowedMediaDomains(std::move(allowedMediaDomains)) {}
+
+absl::Status ImageDecodingProcessor::process(InputRequest& req) {
+    ov::genai::ChatHistory& chatHistory = std::get<ov::genai::ChatHistory>(req.input);
```
Comment (dkalinowski) [inline src/llm/io_processing/input_processors/image_decoding_processor.cpp]: also, do we need try catch and return error?

File: src/llm/io_processing/input_processors/tokenization_processor.hpp [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+// inputIds are used for max-length checks and prompt token usage statistics only;
+// the VLM pipeline tokenizes internally and does not receive inputIds.
+// addSpecialTokens: false for chat path (template already added them), true for completions.
+class TokenizationProcessor : public BaseInputProcessor {
+public:
+    TokenizationProcessor(ov::genai::Tokenizer& tokenizer, bool addSpecialTokens);
+    absl::Status process(InputRequest& req) override;
+
+private:
+    ov::genai::Tokenizer* tokenizer;  // non-owning; lifetime tied to InputProcessorContext
```
Comment (dkalinowski) [inline src/llm/io_processing/input_processors/tokenization_processor.hpp]:
```suggestion
    ov::genai::Tokenizer& tokenizer;  // non-owning; lifetime tied to InputProcessorContext
```

File: src/llm/io_processing/input_processing_config.hpp [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+// Deployment-level configuration for InputProcessor, populated once at servable init.
+struct InputProcessingConfig {
+    // True for VLM servables. Enables ImageDecodingProcessor; TokenizationProcessor
+    // still runs (to populate inputIds for usage statistics and max-length checks)
+    // but inputIds is not passed to the VLM pipeline for inference.
+    bool isVLM = false;
+    // True when the GenAI built-in tokenizer.apply_chat_template() should be used
+    // even on Python-enabled builds (i.e. ChatTemplateMode::MINJA).
+    // False (default) uses PyJinjaTemplateProcessor when PYTHON_DISABLE==0.
+    bool useMinja = false;
```
Comment (dkalinowski) [inline src/llm/io_processing/input_processing_config.hpp:29]: do I put auto deduced caps/workarounds here?

File: src/llm/visual_language_model/continuous_batching/servable.cpp [hunk 8]
```diff
@@ -72,120 +71,4 @@ std::shared_ptr<GenAiServableProperties> VisualLanguageModelServable::getPropert
...
-            }
-            const auto& tools = toolParsingResult.value();
-            auto chatTemplateKwargsParsingResult = vlmExecutionContext->apiHandler->parseChatTemplateKwargsToJsonContainer();
-            if (!chatTemplateKwargsParsingResult.ok()) {
-                return chatTemplateKwargsParsingResult.status();
-            }
-            const auto& chatTemplateKwargs = chatTemplateKwargsParsingResult.value();
-            if (llm_calculator_logger->should_log(spdlog::level::trace)) {
-                SPDLOG_LOGGER_TRACE(llm_calculator_logger, "VLM chatHistory messages: {}", chatHistory.get_messages().to_json_string());
-                SPDLOG_LOGGER_TRACE(llm_calculator_logger, "VLM chatHistory.get_tools(): {}", chatHistory.get_tools().to_json_string());
```
Comment (dkalinowski) [inline src/llm/visual_language_model/continuous_batching/servable.cpp:155]: do we still have such logs somewhere? I think they were useful

File: src/llm/servable.cpp [hunk 9]
```diff
@@ -179,130 +177,59 @@ absl::Status GenAiServable::prepareInputs(std::shared_ptr<GenAiServableExecution
...
-                } catch (const std::exception& e) {
-                    SPDLOG_LOGGER_DEBUG(llm_calculator_logger, "Failed to apply chat template: {}", e.what());
-                    return absl::Status(absl::StatusCode::kInvalidArgument, "Failed to apply chat template. The model either does not have chat template or has an invalid one.");
+    // LM servables reject requests containing image content. Images are preserved
+    // as JsonContainer arrays in chatHistory. Reject if any message's content array
+    // contains an image_url entry.
+    if (!getProperties()->inputProcessorContext.config.isVLM &&
+        std::holds_alternative<ov::genai::ChatHistory>(executionContext->inputRequest.input)) {
+        const auto& ch = std::get<ov::genai::ChatHistory>(executionContext->inputRequest.input);
+        for (size_t i = 0; i < ch.size(); i++) {
```
Comment (dkalinowski) [inline src/llm/servable.cpp:182]: could be separate util function

File: common_settings.bzl [hunk 10]
```diff
@@ -201,6 +201,7 @@ WINDOWS_COMMON_STATIC_LIBS_COPTS = [
                         "/wd6240", 
                         "/wd6326",
                         "/wd6385",
+                        "/wd6386",
```
Comment (dkalinowski) [inline common_settings.bzl:204]: whats this warning?

File: src/test/http_openai_handler_test.cpp [hunk 11]
```diff
@@ -943,97 +849,115 @@ TEST_P(HttpOpenAIHandlerChatAndResponsesParsingTest, ParsingMultipleImagesInOneT
...
+    ASSERT_NE(apiHandler, nullptr);
+    auto content = apiHandler->getChatHistory()[0]["content"];
+    EXPECT_TRUE(content.is_array());
+    EXPECT_EQ(content[0]["type"].as_string().value_or(""), "image_url");
 }
 
-TEST_P(HttpOpenAIHandlerChatAndResponsesParsingTest, ParsingImageStringWithNoMimePrefixFails) {
-    // Without a "data:..." prefix the URL falls through to the local-filesystem loader,
-    // which is disabled by default.
+TEST_P(HttpOpenAIHandlerChatAndResponsesParsingTest, ParsingImageStringWithNoMimePrefixPreservedInChatHistory) {
```
Comment (dkalinowski) [inline src/test/http_openai_handler_test.cpp:877]: if we change the assert here, do we have separate test checking if these requests are rejected during other phase?

File: src/test/http_openai_handler_test.cpp [hunk 12]
```diff
@@ -3281,69 +2437,51 @@ TEST_F(HttpOpenAIHandlerParsingTest, ParsingMessagesWithInvalidContentTypeFails)
...
     std::shared_ptr<ovms::OpenAIChatCompletionsHandler> apiHandler = std::make_shared<ovms::OpenAIChatCompletionsHandler>(doc, ovms::Endpoint::CHAT_COMPLETIONS, std::chrono::system_clock::now(), *tokenizer);
-    EXPECT_EQ(apiHandler->parseMessages(), absl::InvalidArgumentError("Loading images from local filesystem is disabled."));
+    EXPECT_EQ(apiHandler->parseMessages(), absl::InvalidArgumentError("Invalid message structure - content array is empty"));
 }
 
-TEST_F(HttpOpenAIHandlerParsingTest, ParsingMessagesImageUrlNotBase64Fails) {
+TEST_F(HttpOpenAIHandlerParsingTest, ParsingMessagesWithNoContentFieldAddsEmptyStringToChatHistory) {
+    // Assistant turns that carry only tool_calls legitimately omit the "content" field.
+    // parseMessages injects an empty-string content entry so chat templates always see
+    // the field.  Verifies the JsonContainer proxy write-through: lastMessage["content"] = ""
```
Comment (dkalinowski) [inline src/test/http_openai_handler_test.cpp:2347]: this is one of "workarounds" llama.cpp and minja applies. I see we already have it - for all models without autodetection. Is it new behavior or you added this in this PR? If so, which processor handles that?

File: src/llm/apis/openai_api_handler.cpp (hunk 3 above)
Comment (mzegla) [reply src/llm/apis/openai_api_handler.cpp:509]: In new design, the flow makes it impossible (see above req.input = request.chatHistory assignment). 
But we can probably have such defensive try catch here as additional safeguard.

File: src/llm/io_processing/input_processors/image_decoding_processor.cpp (hunk 5 above)
Comment (mzegla) [reply src/llm/io_processing/input_processors/image_decoding_processor.cpp]: Not by design. ImageDecodingProcessor is not supposed to be included for completions endpoint (we don't support VLMs on that endpoint) and for chat and responses endpoints, we get chat history. 
But again, it won't harm to have additional safeguard here, so I can add it.

File: src/llm/io_processing/input_processing_config.hpp (hunk 7 above)
Comment (mzegla) [reply src/llm/io_processing/input_processing_config.hpp:29]: Depends on how they would look like. If they are more like properties (like flags i.e. useMinja = true/false, strings, numbers etc.) that are supposed to be consumed by something - for example input processor reads them and decides how the input processors chain will look like - then yes. 
If they will be more of a fully functional objects themselves (like tokenizer or py jinja template engine) then they should land in the input processing context.

File: src/llm/visual_language_model/continuous_batching/servable.cpp (hunk 8 above)
Comment (mzegla) [reply src/llm/visual_language_model/continuous_batching/servable.cpp:155]: Let's restore them then

File: src/test/http_openai_handler_test.cpp (hunk 11 above)
Comment (mzegla) [reply src/test/http_openai_handler_test.cpp:877]: I'm adding a bunch of new tests with negative scenarios as well. Let's revisit once I push new changes.

File: src/test/http_openai_handler_test.cpp (hunk 12 above)
Comment (mzegla) [reply src/test/http_openai_handler_test.cpp:2347]: I believe we added this a long time ago as some of the models we support for a while now needed to see `content` for every request. Perhaps this was a workaround from the beginning, but at that time, we implemented it as part of core logic.

File: src/llm/io_processing/input_processors/chat_template_processor.cpp [hunk 13]
```diff
@@ -52,11 +53,21 @@ std::string ChatTemplateProcessor::serializeForPyJinja(const ov::genai::ChatHist
...
 #endif
 
 absl::Status ChatTemplateProcessor::process(InputRequest& req) {
-    const ov::genai::ChatHistory& chatHistory = std::get<ov::genai::ChatHistory>(req.input);
+    const auto* chatHistoryPtr = std::get_if<ov::genai::ChatHistory>(&req.input);
+    if (chatHistoryPtr == nullptr) {
+        return absl::Status(absl::StatusCode::kInternal,
+            "ChatTemplateProcessor received input that is not a ChatHistory");
+    }
+    const ov::genai::ChatHistory& chatHistory = *chatHistoryPtr;
```
Comment (dkalinowski) [inline src/llm/io_processing/input_processors/chat_template_processor.cpp]:
```
if (!std::holds_alternative<ov::genai::ChatHistory>(req.input)) {
    return absl::Status(absl::StatusCode::kInternal,
        "ChatTemplateProcessor received input that is not a ChatHistory");
}
const auto& chatHistory = std::get<ov::genai::ChatHistory>(req.input);
```

I think that would be cleaner

File: src/llm/io_processing/input_processors/image_decoding_processor.cpp [hunk 14]
```diff
@@ -31,7 +32,12 @@ ImageDecodingProcessor::ImageDecodingProcessor(
     allowedMediaDomains(std::move(allowedMediaDomains)) {}
 
 absl::Status ImageDecodingProcessor::process(InputRequest& req) {
-    ov::genai::ChatHistory& chatHistory = std::get<ov::genai::ChatHistory>(req.input);
+    auto* chatHistoryPtr = std::get_if<ov::genai::ChatHistory>(&req.input);
+    if (chatHistoryPtr == nullptr) {
+        return absl::Status(absl::StatusCode::kInternal,
+            "ImageDecodingProcessor received input that is not a ChatHistory");
+    }
```
Comment (dkalinowski) [inline src/llm/io_processing/input_processors/image_decoding_processor.cpp:38]: same here

**Review by dkalinowski — APPROVED** [review]

File: common_settings.bzl (hunk 10 above)
Comment (mzegla) [reply common_settings.bzl:204]: Warning from absl. In was there before, but since the code triggering it was located in a larger translation unit, the MSVC static analysis did not go that deep. Now it surfaced as the logic was moved to smaller, more specialized input processor unit.

**Review by pgladkows — APPROVED** [review]
